### PR TITLE
Camp application

### PR DESCRIPTION
**Pull request by zirak-waheed63**

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * It is generally better to create a new file per style scope.
+ *
+ *= require_tree .
+ *= require_self
+ */
+@import "bootstrap-sprockets";
+@import "bootstrap";
+@import 'bootstrap/theme';
+
+.pretty th .current {
```
Comment (john-ibrahim) [inline app/assets/stylesheets/application.scss]: let's create separate file for custom styles

File: app/controllers/admin/camp_applications_controller.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+	private
+
+	def set_application
+		@camp_application = CampApplication.find(params[:application_id])
+	end
+
+	def set_progress
+    @progress = if wizard_steps.any? && wizard_steps.index(step).present?
```
Comment (john-ibrahim) [inline app/controllers/admin/camp_applications_controller.rb]: @progress = (((wizard_steps.index(step) || -1) + 1).to_d / wizard_steps.count.to_d) * 100

File: app/controllers/admin/dashboard_controller.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module Admin
+  class DashboardController < ApplicationController
+    before_action :is_admin?
```
Comment (john-ibrahim) [inline app/controllers/admin/dashboard_controller.rb]: Let's create an Admin::BaseController and  do `is_admin?` in that and inherit every admin controller from that controller

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      u.permit(:first_name, :middle_name, :last_name, :country,
+               :phone_number, :email, :avatar, :password, :password_confirmation, :current_password)
+    end
+  end
+
+  def authenticate_inviter!
+    is_admin?
+  end
+
+  def is_admin?
```
Comment (john-ibrahim) [inline app/controllers/application_controller.rb]: move to `Admin::BaseController`

File: app/controllers/camp_applications_controller.rb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  before_action :validate_dates, only: [:show, :update]
+
+  steps :personal_information, :step_two, :step_three, :step_four, :step_five, :step_six, :step_seven,
+        :step_eight, :step_nine, :step_ten
+
+  def create
+    camp = Camp.find_by(status: true)
+    if camp.nil?
+      redirect_to root_path, alert: 'No active camp available currently' 
+    elsif Date.today > camp.end_date
```
Comment (john-ibrahim) [inline app/controllers/camp_applications_controller.rb]: create a method in camp model `expired?` and move it there

File: app/controllers/camp_applications_controller.rb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  before_action :set_application, except: :create
+  before_action :active?, except: [:view_application, :create]
+  before_action :validate_dates, only: [:show, :update]
+
+  steps :personal_information, :step_two, :step_three, :step_four, :step_five, :step_six, :step_seven,
+        :step_eight, :step_nine, :step_ten
+
+  def create
+    camp = Camp.find_by(status: true)
+    if camp.nil?
```
Comment (john-ibrahim) [inline app/controllers/camp_applications_controller.rb]: `camp.blank?`

File: app/controllers/camp_applications_controller.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    else
+      @camp_application = CampApplication.create(user: current_user, camp: camp)
+      redirect_to dashboard_path
+    end
+    return
+  end
+
+  def show
+    if current_step?(:personal_information)
+      if @camp_application.name.nil?
```
Comment (john-ibrahim) [inline app/controllers/camp_applications_controller.rb]: let's move it in model with method named `initialize_from_user(user)`

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
+class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
```
Comment (john-ibrahim) [inline app/controllers/application_controller.rb]: before_action :authenticate_user!

File: app/controllers/admin/dashboard_controller.rb (hunk 3 above)
Comment (john-ibrahim) [inline app/controllers/admin/dashboard_controller.rb]: Move it  to `Admin::DashboradController < Admin::BaseController`

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  protected
+
+  def configure_permitted_parameters
```
Comment (john-ibrahim) [inline app/controllers/application_controller.rb]: mode to `RegistrationsController < Devise::RegistrationsController`

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    devise_parameter_sanitizer.permit(:invite,
+                                      keys: %i[first_name middle_name last_name phone_number country email])
+    devise_parameter_sanitizer.permit(:account_update) do |u|
+      u.permit(:first_name, :middle_name, :last_name, :country,
+               :phone_number, :email, :avatar, :password, :password_confirmation, :current_password)
+    end
+  end
+
+  def authenticate_inviter!
+    is_admin?
```
Comment (john-ibrahim) [inline app/controllers/application_controller.rb]: change as discussed, move to InvitationController, update definition

File: app/controllers/camp_applications_controller.rb (hunk 6 above)
Comment (john-ibrahim) [inline app/controllers/camp_applications_controller.rb]: replace `.nil?` with `.blank?` throughout the application

File: app/controllers/camp_applications_controller.rb (hunk 7 above)
Comment (john-ibrahim) [reply app/controllers/camp_applications_controller.rb]: `name = "#{user.first_name} #{user.middle_name} #{user.last_name}" if name.blank?`

File: app/controllers/camp_applications_controller.rb [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      @camp_application.email = current_user.email if @camp_application.email.nil?
+      if !@camp_application.image.attached? && current_user.avatar.attached?
+        @camp_application.image.attach(current_user.avatar.blob)
+      end
+    end
+    render_wizard
+  end
+
+  def update
+    @camp_application.update_attributes(application_params)
```
Comment (john-ibrahim) [inline app/controllers/camp_applications_controller.rb]: if @camp_application.update(application_params)
  @camp_application.update_progress
  @camp_application.update_status
else
  render
end

File: app/controllers/camp_applications_controller.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
+class CampApplicationsController < Wicked::WizardController
+  before_action :not_admin?
```
Comment (adam-sheikh1) [inline app/controllers/camp_applications_controller.rb]: move it to `Admin::BaseController`?

File: app/controllers/camp_applications_controller.rb [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    elsif camp.is_finished?
+      redirect_to root_path, alert: 'Camp has ended. Please participate in next camp'
+    else
+      @camp_application = CampApplication.create(user: current_user, camp: camp)
+      redirect_to dashboard_path
+    end
+  end
+
+  def show
+
```
Comment (adam-sheikh1) [inline app/controllers/camp_applications_controller.rb]: remove blank line

File: app/controllers/camp_locations_controller.rb [hunk 14]
```diff
@@ -0,0 +1,60 @@
+class CampLocationsController < Admin::BaseController
+  before_action :set_camp_location, only: %i[ show edit update destroy ]
```
Comment (adam-sheikh1) [inline app/controllers/camp_locations_controller.rb]: Do it as:
```suggestion
  before_action :set_camp_location, only: %i[show edit update destroy]
```

File: app/views/admin/camp_applications/step_ten.html.erb [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+<%= form_for @camp_application, url: wizard_path, method: :put do |f| %>
+  <div class="field">
+      <%= f.label :step_ten %><br />
+      <%= f.text_field :step_ten %>
+    </div>
+
+    <div class="actions">
+      <%= f.submit "Submit" %>
+    </div>
+<%end %>
```
Comment (adam-sheikh1) [inline app/views/admin/camp_applications/step_ten.html.erb]: `<% end %>`

Also fix indentation in this file

File: app/views/admin/camp_applications/step_three.html.erb [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+<%= form_for @camp_application, url: wizard_path, method: :put do |f| %>
+  <div class="field">
+      <%= f.label :step_three %><br />
+      <%= f.text_field :step_three %>
+    </div>
+
+    <div class="actions">
+      <%= f.submit "done"%>
+    </div>
+<%end %>
```
Comment (adam-sheikh1) [inline app/views/admin/camp_applications/step_three.html.erb]: `<% end %>`

Also fix indentation in this file

File: app/views/admin/camp_applications/step_two.html.erb [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+<%= form_for @camp_application, url: wizard_path, method: :put do |f| %>
+  <div class="field">
+      <%= f.label :step_two %><br />
+      <%= f.text_field :step_two %>
+    </div>
+
+    <div class="actions">
+      <%= f.submit "done"%>
+    </div>
+<%end %>
```
Comment (adam-sheikh1) [inline app/views/admin/camp_applications/step_two.html.erb]: `<% end %>`

Also fix indentation in this file

File: app/controllers/camp_locations_controller.rb [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+  end
+
+  def show
+  end
+
+  def new
+    @camp_location = CampLocation.new
+  end
+
+  def edit
```
Comment (adam-sheikh1) [inline app/controllers/camp_locations_controller.rb]: def edit; end

File: app/controllers/pages_controller.rb [hunk 19]
```diff
@@ -0,0 +1,26 @@
+class PagesController < ApplicationController
+  before_action :not_admin?
```
Comment (adam-sheikh1) [inline app/controllers/pages_controller.rb]: let's make a base controller for user as well

File: app/controllers/pages_controller.rb [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+
+  def introduction
+    redirect_to dashboard_path if current_user.camp_application.present?
+    @camp = Camp.find_by(status: true)
+    flash[:alert] = 'No active camp available currently' if @camp.blank?
+  end
+
+  def dashboard
+    @camp_application = current_user.camp_application
+    if @camp_application.blank?
```
Comment (adam-sheikh1) [inline app/controllers/pages_controller.rb]: `return redirect_to root_path if @camp_application.blank?`

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,77 @@
+class UsersController < Admin::BaseController
+  helper_method :sort_column, :sort_direction
+  before_action :check_logged_in, only: :profile
+  skip_before_action :is_admin?, only: :profile
+  def profile
```
Comment (adam-sheikh1) [inline app/controllers/users_controller.rb:2]: add an empty line above this

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -0,0 +1,77 @@
+class UsersController < Admin::BaseController
```
Comment (adam-sheikh1) [inline app/controllers/users_controller.rb]: let's create two users controller
Admin::UsersController
UsersController

File: app/models/user.rb [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+    }
+    if password.present?
+      rules.each do |message, regex|
+        errors.add( :password, message ) unless password.match(regex)
+      end
+    end
+  end
+
+  def self.search(query)
+    where('first_name LIKE ? OR last_name LIKE ? OR email LIKE ? OR id LIKE ?',
```
Comment (adam-sheikh1) [inline app/models/user.rb]: `return all if query.blank?`

File: app/controllers/users_controller.rb [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+  helper_method :sort_column, :sort_direction
+  before_action :check_logged_in, only: :profile
+  skip_before_action :is_admin?, only: :profile
+  def profile
+    @user = current_user
+  end
+
+  def index
+    @users = if params[:search].present?
```
Comment (adam-sheikh1) [inline app/controllers/users_controller.rb]: @search = params[:search] || ''
@users = User.search(@search).custom_order(sort_column, sort_direction).page(params[:page])

File: app/controllers/users_controller.rb [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
```
Comment (adam-sheikh1) [inline app/controllers/users_controller.rb]: refactor this, remove password fields

File: app/helpers/application_helper.rb [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+  def header(text)
+    content_for(:header) { text.to_s }
+  end
+
+  def sort_link(column, title = nil)
+    title ||= column.titleize
+    direction = column == sort_column && sort_direction == "asc" ? "desc" : "asc"
+    icon = sort_direction == "asc" ? "glyphicon glyphicon-chevron-up" : "glyphicon glyphicon-chevron-down"
+    icon = column == sort_column ? icon : ""
+    link_to "#{title} <span class='#{icon}'></span>".html_safe, {column: column, direction: direction}
```
Comment (adam-sheikh1) [inline app/helpers/application_helper.rb]:
```suggestion
    link_to "#{title} <span class='#{icon}'></span>".html_safe, { column: column, direction: direction }
```
[] donot add spaces at start/end
() donot add spaces at start/end
{} add spaces at start/end

File: app/helpers/camp_applications_helper.rb [hunk 27]
```diff
@@ -0,0 +1,8 @@
+module CampApplicationsHelper
+	def show_confirmation(application, status)
+		if application.progress.equal?(100) || (application.progress.equal?(90) && application.step_ten.blank?)
+			arr = ['confirm', 'You cannot edit application after submission. Are you sure to submit?']
```
Comment (adam-sheikh1) [inline app/helpers/camp_applications_helper.rb]: { 'confrim' => 'You cannot edit application after submission. Are you sure to submit?' }

File: app/models/camp_application.rb [hunk 28]
```diff
@@ -0,0 +1,39 @@
+class CampApplication < ApplicationRecord
+  belongs_to :camp
+  belongs_to :user
+  has_one_attached :image
+  after_initialize :initialize_steps_hash
+  validates :name, length: {minimum: 5}, on: :update, allow_nil: true
+  
+  STEPS = %w[personal_information step_two step_three step_four step_five step_six step_seven step_eight step_nine step_ten]
```
Comment (adam-sheikh1) [inline app/models/camp_application.rb:8]: add empty line below

File: app/models/camp_application.rb [hunk 29]
```diff
@@ -0,0 +1,39 @@
...
+  def complete_step(name)
+    current_steps = steps
+    current_steps[name] = true
+    update(steps: current_steps)
+  end
+
+  def status
+    status = steps.select { |k, v| !v}.keys.first
+    status = 'active' if status.blank?
+    return status
```
Comment (adam-sheikh1) [inline app/models/camp_application.rb]:
```suggestion
    status
```

File: app/models/camp_application.rb [hunk 30]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def progress
+    completed = steps.select { |k, v| v }.keys.size
+    total = steps.keys.size
+    return completed*100/total
```
Comment (adam-sheikh1) [inline app/models/camp_application.rb]:
```suggestion
    completed*100/total
```

File: app/models/camp_application.rb [hunk 31]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def initialize_from_user(user)
+    self.name = "#{user.first_name} #{user.middle_name} #{user.last_name}" if name.blank?
```
Comment (adam-sheikh1) [inline app/models/camp_application.rb]: self.name = user.full_name

in user model:

def full_name
    [first_name, middle_name, last_name].reject(&:blank?).join(' ')
end

Comment (adam-sheikh1) [inline app/models/camp_application.rb]: return if persisted?

File: app/models/user.rb [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+                    length: { minimum: 4, maximum: 254 }                                                        
+  validate :password_requirements_are_met
+
+  has_one :camp_application, dependent: :destroy
+  has_one_attached :avatar
+  
+  devise :invitable, :database_authenticatable, :registerable, :confirmable,
+         :recoverable, :rememberable, :validatable
+
+  private
```
Comment (adam-sheikh1) [inline app/models/user.rb:23]: add empty line

File: app/models/user.rb (hunk 23 above)
Comment (adam-sheikh1) [inline app/models/user.rb]: ILIKE

File: app/views/admin/camp_applications/_progressbar.html.erb [hunk 33]
```diff
@@ -0,0 +1,23 @@
+<div id="msform">
+  <ul id="progressbar">
+    <li class= "<%='active' if @camp_application.steps['personal_information'].present?%> <%= 'current' if current_step == 'personal_information'%>" ><%= link_to 'Personal Information', admin_camp_application_path(application_id: @camp_application.id, id: 'personal_information') %></li>
```
Comment (adam-sheikh1) [inline app/views/admin/camp_applications/_progressbar.html.erb]: fix indentation
Add empty line at end of file
let's move this logic to a loop

File: app/views/admin/dashboard/index.html.erb [hunk 34]
```diff
@@ -0,0 +1,16 @@
+<div class="jumbotron">
+  <div class="container">
+    <h1>Welcome!</h1>
+    <%if current_user.present? && current_user.admin? %>
+      <%= link_to 'List all Users', users_path, class: 'btn btn-default' %>
```
Comment (adam-sheikh1) [inline app/views/admin/dashboard/index.html.erb]: <'div class="mb-2">
      <%= link_to 'List all Users', users_path, class: 'btn btn-default' %>
<'/div>

File: app/views/users/index.html.erb [hunk 35]
```diff
@@ -0,0 +1,44 @@
...
+        <td class="col col-md-2"><%= country_name(user.country) %></td>
+        <td class="col col-md-3"><%= user.phone_number %></td>
+        <td><%= link_to 'Show', user %></td>
+        <td><%= link_to 'Edit', edit_user_path(user) %></td>
+        <td><%= link_to 'Destroy', user, method: :delete, data: { confirm: 'Are you sure?' } %></td>
+      </tr>
+    <% end %>
+  </tbody>
+</table>
+<%=paginate @users %>
```
Comment (adam-sheikh1) [inline app/views/users/index.html.erb]:
```suggestion
<%= paginate @users %>
```

File: db/migrate/20210225103340_add_confirmable_to_devise.rb [hunk 36]
```diff
@@ -0,0 +1,10 @@
+class AddConfirmableToDevise < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :confirmation_token, :string
+    add_column :users, :confirmed_at,       :datetime
+    add_column :users, :confirmation_sent_at , :datetime
+    add_column :users, :unconfirmed_email, :string
+
+    add_index  :users, :confirmation_token, :unique => true
```
Comment (adam-sheikh1) [inline db/migrate/20210225103340_add_confirmable_to_devise.rb]:
```suggestion
    add_index  :users, :confirmation_token, unique: true
```

File: app/assets/javascripts/custom.js [hunk 37]
```diff
@@ -0,0 +1,21 @@
...
+    if(this.checked == true){
+      document.getElementById('overseas-fields').classList.remove('hidden')
+      }
+    else{
+      document.getElementById('overseas-fields').classList.add('hidden')
+    }
+  });
+
+  $('#officer').change(function() {
+    if(this.checked == true){
```
Comment (john-ibrahim) [inline app/assets/javascripts/custom.js]:
```suggestion
    if (this.checked == true) {
```

File: app/assets/javascripts/custom.js [hunk 38]
```diff
@@ -0,0 +1,21 @@
...
+      document.getElementById('officers-fields').classList.remove('hidden')
+      }
+    else{
```
Comment (john-ibrahim) [inline app/assets/javascripts/custom.js]:
```suggestion
    else {
```

File: app/controllers/camp_locations_controller.rb [hunk 39]
```diff
@@ -0,0 +1,59 @@
...
+    @camp_location.destroy
+    respond_to do |format|
+      format.html { redirect_to camp_locations_url, notice: "Camp location was successfully destroyed." }
+      format.json { head :no_content }
+    end
+  end
+
+  private
+  
+    def set_camp_location
```
Comment (john-ibrahim) [inline app/controllers/camp_locations_controller.rb]: donot add extra tab after private blocks

File: app/models/user.rb [hunk 40]
```diff
@@ -0,0 +1,53 @@
...
+      rules.each do |message, regex|
+        errors.add(:password, message) unless password.match(regex)
+      end
+    end
+  end
+
+  def self.search(query)
+    return all if query.blank?
+    where('first_name LIKE ? OR last_name LIKE ? OR email LIKE ? OR id LIKE ?',
+          "%#{ query }%", "%#{ query }%", "%#{ query }%", "%#{ query }%")
```
Comment (john-ibrahim) [inline app/models/user.rb]: sorry, we don't add  extra spaces in curly brackets while interpolating string, kindly update

File: app/views/admin/users/edit.html.erb [hunk 41]
```diff
@@ -0,0 +1,52 @@
+<h2>Edit User</h2>
+  <ul>
```
Comment (john-ibrahim) [inline app/views/admin/users/edit.html.erb]: indentation?

File: app/views/admin/users/edit.html.erb [hunk 42]
```diff
@@ -0,0 +1,52 @@
...
+    <% @user.errors.full_messages.each do |error| %>
+      <li><%= error %></li>
+    <% end %>
+  </ul>
+<%= form_for @user, url: admin_user_path, method: :put do |f| %>
+    <div>
```
Comment (john-ibrahim) [inline app/views/admin/users/edit.html.erb]: incorrect indentation, please fix

File: app/views/admin/users/index.html.erb [hunk 43]
```diff
@@ -0,0 +1,43 @@
+<h1>All Users</h1>
+  <div>
```
Comment (john-ibrahim) [inline app/views/admin/users/index.html.erb]: indentation?

File: app/views/admin/users/index.html.erb [hunk 44]
```diff
@@ -0,0 +1,43 @@
...
+      <th class="col col-md-3"><%= sort_link "phone_number" %></th>
+      <th colspan="3"></th>
+    </tr>
+  </thead>
+
+  <tbody>
+    <% @users.each do |user| %>
+      <tr class="row">
+        <td class="col col-md-3">
+          <%= user.first_name + ' ' %>
```
Comment (john-ibrahim) [inline app/views/admin/users/index.html.erb]: can we use `user.full_name`?

File: app/views/admin/users/show.html.erb [hunk 45]
```diff
@@ -0,0 +1,25 @@
+<h2>Show User Profile</h2>
+
+<div class="container">
+
+  <div>
```
Comment (john-ibrahim) [inline app/views/admin/users/show.html.erb:4]: indentation?

File: app/views/admin/users/show.html.erb [hunk 46]
```diff
@@ -0,0 +1,25 @@
...
+    <% if @user.avatar.attached? %>
+      <img src="<%= url_for(@user.avatar) %>" class="img-circle img-responsive" style="height: 200px; width: 200px">
+    <% end %>
+  </div>
+  <div class="field">
+    <b>Name:</b>
+    <%= @user.first_name + ' ' %>
```
Comment (john-ibrahim) [inline app/views/admin/users/show.html.erb]: `user.full_name`

**Pull request closed without merging**